### DISCUSSION TITLE
Vault Rebalance

### DISCUSSION THREAD

**Discussion by brozorec** [Fujicracy/fuji-v2#50] [General]
@DaigaroCota 
We haven't talked yet about how to do the rebalances of the vaults. In v1, we had a dedicated function in the vault. Do we have better ideas on how to do it in v2?

It's worth considering how are we going to trigger those functions. Are we going to use OZ Defender or something else (Keep3r, Gelato, ..)?

I found this to integrate Keep3r with Defender and like this, we won't need to run some scripts on our servers:
https://docs.openzeppelin.com/defender/guide-keep3r

**Comment by 0xdcota** [comment]
I am ok on using Keep3rs from the Defender platform.

In terms of the rebalance, I think we should probably stick to having the logic in the vault contract. This with the purpose of avoiding more external contracts (which increase attack vector surface).

I propose we call the function `rebalance` and that the arguments are:
- newProvider: function must validate this is a valid provider.
- assetAmount:
- debtAmount: this should be zero for a yield  vault
- flashloanProviderId

The function `rebalance` should return a bool at the end if all succeeded. 

The following specs will apply too:

- Rebalance will need to know the address of the {Flasher} contract, or get it from a contract who know who is the Flasher
- A specific role at some hierarchy level will need to be defined as "rebalancer"
- The callback detailed logic will have to live in the Vault contract. There will probably be required two callback logics. One for yield vaults and one for borrowing vaults.

**Reply by brozorec** [reply]
Regarding `flashloanProviderId`, what about we don't refer to them with ids but rather with addresses?

> Rebalance will need to know the address of the {Flasher} contract, or get it from a contract who know who is the Flasher

I suggest we keep a reference to allowed Flashers in `Chief.sol`.

> A specific role at some hierarchy level will need to be defined as "rebalancer"

I agree with this. If we go with this, we first need to implement the roles in the Vault before proceeding with rebalancing.

What about if the rebalancer is another contract that handles the flashloan initiation and callbacks so that we don't supercharge the vault contract?

**Reply by 0xdcota** [reply]
Yes I think using addresses will be better. Also the control at the Chief level sounds good, and perhaps also a priority as you suggest. In general there is many other access control items that we need to determine and define to control at Chief level.  Some things that come to my mind:

- Liquidator role
- Rebalancer role
- Pauser role (for emergencies, this should be some type of Defender Sentinel?)
- unpauser role
- general admin and Risk parameter changing role (address setter, maxLTV values, discounted liquidation price rate, future fees?)

**Reply by brozorec** [reply]
here is the breakdown of access control and roles:
https://github.com/Fujicracy/fuji-v2/issues/78#issuecomment-1274413163
